File: 04256.cpp
#include <iostream>

using namespace std;

int search(int arr[], int x, int n){
	for(int i = 0; i < n; ++i){
		if(arr[i] == x){
			return i;
		}
	}
	return -1;
}

void postorder(int pre[], int in[], int n){
	int root = search(in, pre[0], n);
	if(root != 0){
		postorder(pre + 1, in, root);
	}
	if(root != n - 1){
		postorder(pre + root + 1, in + root + 1, n - 1 - root);
	}
	cout << pre[0] << ' ';
}

int main(){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	int t;
	cin >> t;
	for(int a0 = 0; a0 < t; ++a0){
		int n;
		cin >> n;
		int preorder[1001], inorder[1001];
		for(int i = 0; i < n; ++i){
			cin >> preorder[i];
		}
		for(int i = 0; i < n; ++i){
			cin >> inorder[i];
		}
		postorder(preorder, inorder, n);
		cout << '\n';
	}
	return 0;
}

File: 04256.py
import sys


def postorder(preorder, inorder, n):
    ret = []
    root = inorder.index(preorder[0])
    if root != 0:
        ret += postorder(preorder[1:], inorder, root)
    if root != n - 1:
        ret += postorder(preorder[1 + root:], inorder[1 + root:], n - root - 1)
    ret.append(preorder[0])
    return ret


def main():
    t = int(sys.stdin.readline())
    for _ in range(t):
        n = int(sys.stdin.readline())
        preorder = [int(x) for x in sys.stdin.readline().split()]
        inorder = [int(x) for x in sys.stdin.readline().split()]
        print(" ".join(str(x) for x in postorder(preorder, inorder, n)))


main()
